File: leetcode/algorithms/2_add_two_numbers/main.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class AddTwoNumbers:
    """
    # Complexities:
    #   M - l1 length
    #   N - l2 length
    #   - Time Complexity: O(Max(M, N))
    #   - Space Complexity: O(Max(M, N))
    """
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry: int = 0

        result: Optional[ListNode] = ListNode()
        head: Optional[ListNode] = result

        while l1 != None or l2 != None or carry > 0:
            sum: int = carry
            carry = 0

            if l1 != None:
                sum += l1.val
                l1 = l1.next

            if l2 != None:
                sum += l2.val
                l2 = l2.next

            if sum > 9:
                sum -= 10
                carry = 1

            head.next = ListNode(sum)
            head = head.next

        return result.next


    # Solution
    """
    # Complexities:
    #   M - l1 length
    #   N - l2 length
    #   - Time Complexity: O(Max(M, N))
    #   - Space Complexity: O(Max(M, N))
    """
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        current = dummy
        carry = 0

        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0

            total = val1 + val2 + carry

            carry = total // 10
            digit = total % 10

            current.next = ListNode(digit)
            current = current.next

            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        return dummy.next

File: leetcode/algorithms/2_add_two_numbers/main.java
class AddTwoNumbers {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    /**
     * Complexities:
     *   M - Length of `l1`
     *   N - Length of `l2`
     *   - Time Complexity: O(Max(M, N))
     *   - Space Complexity: O(Max(M, N))
     */
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;

        ListNode result = new ListNode();
        ListNode head = result;

        while (l1 != null || l2 != null || carry > 0) {
            int sum = carry;
            carry = 0;

            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }

            if (sum > 9) {
                sum -= 10;
                carry = 1;
            }

            head.next = new ListNode(sum);
            head = head.next;
        }

        return result.next;
    }


    // Solution
    /**
     * Complexities:
     *   M - Length of `l1`
     *   N - Length of `l2`
     *   - Time Complexity: O(Max(M, N))
     *   - Space Complexity: O(Max(M, N))
     */
    public ListNode solution(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode curr = dummyHead;

        int carry = 0;

        while (l1 != null || l2 != null || carry > 0) {
            int x = (l1 != null) ? l1.val : 0;
            int y = (l2 != null) ? l2.val : 0;

            int sum = carry + x + y;

            carry = sum / 10;

            curr.next = new ListNode(sum % 10);

            curr = curr.next;
            if (l1 != null) {
                l1 = l1.next;
            }
            if (l2 != null) {
                l2 = l2.next;
            }
        }

        return dummyHead.next;
    }
}
